tests/FsCodec.SystemTextJson.Tests/UnionConverterTests.fs

```diff
@@ -0,0 +1,361 @@
+﻿module FsCodec.SystemTextJson.Tests.UnionConverterTests
+
+open FsCheck
+open FsCodec.SystemTextJson
+open Swensen.Unquote.Assertions
+open System
+open System.IO
+open System.Text.Json
+open System.Text.Json.Serialization
+open global.Xunit
+
+open Samples
+
+// TODO support [<Struct>]
+type TestRecordPayload =
+    {
+        test: string
+    }
+
+// TODO support [<Struct>]
+type TrickyRecordPayload =
+    {
+        Item: string
+    }
+
+[<JsonConverter(typeof<TypeSafeEnumConverter<Mode>>)>]
+type Mode =
+    | Fast
+    | Slow
+
+[<NoComparison>] // NB this is not a general restriction; it's forced by use of Nullable<T> in some of the cases in this specific one
+[<JsonConverter(typeof<UnionConverter<TestDU>>)>]
+type TestDU =
+    | CaseA of TestRecordPayload
+    | CaseB
+    | CaseC of string
+    | CaseD of a: string
+    | CaseE of string * int
+    | CaseF of a: string * b: int
+    | CaseG of TrickyRecordPayload
+    | CaseH of a: TestRecordPayload
+    | CaseI of a: TestRecordPayload * b: string
+    | CaseJ of a: Nullable<int>
+    | CaseK of a: int * b: Nullable<int>
+    | CaseL of a: Nullable<int> * b: Nullable<int>
+    | CaseM of a: int option
+    | CaseN of a: int * b: int option
+    | CaseO of a: int option * b: int option
+    | CaseP of CartId
+    | CaseQ of SkuId
+    | CaseR of a: CartId
+    | CaseS of a: SkuId
+    | CaseT of a: SkuId option * b: CartId
+    | CaseU of SkuId[]
+    | CaseV of skus: SkuId[]
+    | CaseW of CartId * SkuId[]
+    | CaseX of a: CartId * skus: SkuId[]
+    | CaseY of a: Mode * b: Mode
+    | CaseZ of a: Mode * b: Mode option
+
+#nowarn "1182" // From hereon in, we may have some 'unused' privates (the tests)
+
+// no camel case, because I want to test "Item" as a record property
+// Centred on ignoreNulls for backcompat; round-tripping test covers the case where they get rendered too
+let defaultOptions = Options.Create(converters = [| UnionConverter<TestDU> () |], camelCase = false, ignoreNulls = true)
```

Comment:
TODO:  Converter not being picked for serialization from attribute, only when specified in JsonSerializationOptions